app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,314 @@
+package de.westnordost.streetcomplete.data.osm.mapdata
+
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometry
+import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometryEntry
+import de.westnordost.streetcomplete.data.osm.geometry.ElementPointGeometry
+import de.westnordost.streetcomplete.util.SpatialCache
+
+/**
+ * Cache for MapDataController using SpatialCache for nodes
+ * caches data returned by db, except for nodes
+ * for tiles in spatial cache all elements and geometries are cached,
+ * way and relation data outside the cached tiles may be cached, but will be removed on any trim
+ */
+class MapDataCache(
+    private val tileZoom: Int,
+    maxTiles: Int,
+    initialCapacity: Int,
+    val fetchMapData: (BoundingBox) -> Pair<Collection<Element>, Collection<ElementGeometryEntry>>, // used if the tile is not contained
+) {
+    private val spatialCache = SpatialCache(
+        tileZoom,
+        maxTiles,
+        initialCapacity,
+        { emptyList() }, // data is fetched using fetchMapData and put using spatialCache.replaceAllInBBox
+        Node::id, Node::position
+    )
+    // initial values obtained from a spot check:
+    //  approximately 80% of all elements were found to be nodes
+    //  approximately every second node is part of a way
+    //  more than 90% of elements are not part of a relation
+    private val wayRelationCache = HashMap<ElementKey, Element?>(initialCapacity / 6)
+    private val wayRelationGeometryCache = HashMap<ElementKey, ElementGeometry?>(initialCapacity / 6)
```

Comment:
I had tried that earlier, and found that the only performance difference was that `getMapDataWithGeometry` was faster when using the `wayRelationCache`.
Though that was an earlier version... I'll re-test that later and switch if there is no benefit the current way

Reply:
Sure, we can leave this discussion open for now and revisit it later.

Reply:
See: https://github.com/Helium314/StreetComplete/compare/cache...Helium314-cache-patch-1
Tested again, and this time no performance differences found.

I don't have any preference what to use.

Reply:
Oh, that's nice! I think I did (slightly?) prefer to have it separate even if there was no (measurable) performance difference because it more closely mirrors how the database is structured (ways and relations are each in separate tables).

Reply:
sorry, I must have overlooked the notification...
done now